### PR TITLE
Removing S3's old Signature; use Signature Version 4 exclusively

### PR DESCRIPTION
**Pull request by jeremeamia**
* Removes `Aws\Signature\S3Signature` and related code and tests.
* Removes `'force_path_style'` S3 client option, because path_style will not be used exclusively.
* Removes `'calculate_md5'` option, because Signature Version 4 already does sha256 calculations.
* Simplifies several middleware and the multipart uploader due to these changes.
* Fixes issues with `LocationConstraint` in `GetBucketLocation`.

### CODE REVIEW

File: src/S3/BucketStyleMiddleware.php [hunk 1]
```diff
@@ -15,28 +15,22 @@
 class BucketStyleMiddleware
 {
     private static $exclusions = ['GetBucketLocation' => true];
-    private $bucketEndpoint;
```
Comment (mtdowling) [inline src/S3/BucketStyleMiddleware.php]: Why was this removed? Did it go somewhere else?

**Comment by mtdowling** [comment]
I like it, but can you tell me how `$bucketEndpoint` is implemented?

File: src/S3/BucketStyleMiddleware.php (hunk 1 above)
Comment (jeremeamia) [inline src/S3/BucketStyleMiddleware.php]: Not needed because this middleware's sole reason to exist now is to handle bucket endpoints.

**Comment by jeremeamia** [comment]
BucketStyle middleware only gets added if bucket_endpoint config is specified: https://github.com/aws/aws-sdk-php/pull/581/files#diff-4ca0826aef3c8f72db542dfae087e933R70

**Comment by mtdowling** [comment]
Ah ok. If that's all it does, what would you think about renaming it to `BuckeEndpointMiddleware`? Other than that, :ship:

**Comment by mtdowling** [comment]
:ship:
